pkg/utils/tsoutil/tso_dispatcher.go

```diff
@@ -75,18 +75,18 @@ func (s *TSODispatcher) DispatchRequest(serverCtx context.Context, req Request,
 	key := req.getForwardedHost()
 	val, loaded := s.dispatchChs.Load(key)
 	if !loaded {
-		val = &tsoRequestProxyQueue{requestCh: make(chan Request, maxMergeRequests+1)}
-		val, loaded = s.dispatchChs.LoadOrStore(key, val)
+		tsoQueue := &tsoRequestProxyQueue{requestCh: make(chan Request, maxMergeRequests+1)}
+		dispatcherCtx, ctxCancel := context.WithCancelCause(serverCtx)
+		tsoQueue.ctx = dispatcherCtx
+		tsoQueue.cancel = ctxCancel
+		val, loaded = s.dispatchChs.LoadOrStore(key, tsoQueue)
 	}
 	tsoQueue := val.(*tsoRequestProxyQueue)
 	if !loaded {
 		log.Info("start new tso proxy dispatcher", zap.String("forwarded-host", req.getForwardedHost()))
 		tsDeadlineCh := make(chan *TSDeadline, 1)
-		dispatcherCtx, ctxCancel := context.WithCancelCause(serverCtx)
-		tsoQueue.ctx = dispatcherCtx
-		tsoQueue.cancel = ctxCancel
 		go s.dispatch(tsoQueue, tsoProtoFactory, req.getForwardedHost(), req.getClientConn(), tsDeadlineCh, tsoPrimaryWatchers...)
-		go WatchTSDeadline(dispatcherCtx, tsDeadlineCh)
+		go WatchTSDeadline(tsoQueue.ctx, tsDeadlineCh)
 	}
```

Comment:
The current logic is somewhat confusing. I recommend moving this `if` statement to immediately after line 82. This will clarify that it is only triggered when the `LoadOrStore` executes successfully.